src/nvim/ex_cmds.c

```diff
@@ -6156,7 +6249,11 @@ void ex_substitute(exarg_T *eap)
   curwin->w_p_cul = false;    // Disable 'cursorline'
   curwin->w_p_cuc = false;    // Disable 'cursorcolumn'
 
+  long numval = 0;
+  get_option_value((char_u *)"hls", &numval, NULL, 0);
+  set_option_value((const char *)"hls", 0, NULL, OPT_GLOBAL);
   buf_T *preview_buf = do_sub(eap, profile_setlimit(p_rdt));
+  set_option_value((const char *)"hls", numval, NULL, OPT_GLOBAL);
```

Comment:
If possible, it's often faster to set/reset options directly (e.g. `p_hls` in this case), particuarly if there's side-effects. I usually look around the codebase to see how it's typically done for a given option (didn't check `p_hls`), some options will behave badly if their side-effects aren't taken care of. 

In this case I would _guess_ that `p_hls = false` is enough to get what we want, without any problems. Maybe you tried it already though.

Reply:
I did neither try nor think of it :) I'll give it a a shot, but serious consideration will have to wait until the weekend.